syntax = "proto3";
option cc_enable_arenas = true;

package NFq.NConfig;
option java_package = "ru.yandex.kikimr.proto";

import "ydb/core/fq/libs/config/protos/storage.proto";

import "ydb/library/yql/dq/actors/protos/dq_status_codes.proto";
import "ydb/public/api/protos/ydb_query.proto";

////////////////////////////////////////////////////////////

message TMapItem {
    string Key = 1;
    string Value = 2;
}

message TQueryMapping {
    repeated TMapItem CloudIdToTenantName  = 1;
    repeated TMapItem ScopeToTenantName    = 2;
    repeated string CommonTenantName       = 3;
}

// 1. StatusCode(s) are handled with defined policies, non-unique StatusCode(s) across all policies is UB
// 2. RetryCount and RetryPeriodMs are used to calculate actual RetryRate, if it exceeds RetryCount, query is aborted
// 3. BackoffPeriodMs is factor of RetryRate to delay query execution before next retry
// 4. There are no default retry policy, all unhandled statuses are fatal

message TRetryPolicy {
    uint64 RetryCount = 1;
    string RetryPeriod = 2;
    string BackoffPeriod = 3;
}

message TRetryPolicyMapping {
    repeated NYql.NDqProto.StatusIds.StatusCode StatusCode = 1;
    TRetryPolicy Policy = 2;
}

message TControlPlaneStorageConfig {
    bool Enabled = 1;
    NFq.NConfig.TYdbStorageConfig Storage = 2; // TODO: remove
    string IdempotencyKeysTtl = 3;
    uint64 MaxRequestSize = 4;
    uint64 MaxCountQueries = 5;
    uint64 MaxCountConnections = 6;
    uint64 MaxCountBindings = 7;
    uint64 MaxCountJobs = 8;
    repeated string SuperUsers = 9;
    bool EnableDebugMode = 10;
    bool EnablePermissions = 11;
    bool DisableCurrentIam = 12;
    bool UseInMemory = 13;
    uint64 TasksBatchSize = 14;
    uint64 NumTasksProportion  = 15;
    uint64 AnalyticsRetryCounterLimit = 16;      // DEPRECATED by RetryPolicy::RetryCount
    uint64 StreamingRetryCounterLimit = 17;      // DEPRECATED by RetryPolicy::RetryCount
    string AnalyticsRetryCounterUpdateTime = 18; // DEPRECATED by RetryPolicy::RetryPeriodMs
    string StreamingRetryCounterUpdateTime = 19; // DEPRECATED by RetryPolicy::RetryPeriodMs
    string AutomaticQueriesTtl = 20;
    string TaskLeaseTtl = 21;
    TRetryPolicy TaskLeaseRetryPolicy = 27;
    repeated string AvailableConnection = 22;
    repeated string AvailableBinding = 23;
    string ResultSetsTtl = 24;
    TQueryMapping Mapping = 25;
    repeated TRetryPolicyMapping RetryPolicyMapping = 26;
    string QuotaTtl = 28;
    string MetricsTtl = 29;
    bool UseDbMapping = 30;
    string DbReloadPeriod = 31;
    bool DumpRawStatistics = 32;
    bool IgnorePrivateSources = 33;
    Ydb.Query.StatsMode StatsMode = 34;
    repeated string AvailableStreamingConnection = 35;
}
